src/USER-MDI/mdi_engine.cpp

```diff
@@ -0,0 +1,364 @@
+
+  // the MD simulation is now at the @INIT_MD node
+  char *command = NULL;
+  command = mdi_fix->engine_mode("@INIT_MD");
+
+  if (strcmp(command,"@DEFAULT") == 0 ) {
+    // return, and flag for @DEFAULT node
+    return 0;
+  }
+  else if (strcmp(command,"EXIT") == 0 ) {
+    // return, and flag for global exit
+    return 1;
+  }
+
+  // continue the MD simulation
+  update->integrate->setup(1);
+
+  // the MD simulation is now at the @FORCES node
+  command = mdi_fix->engine_mode("@FORCES");
+
+  if (strcmp(command,"@DEFAULT") == 0 ) {
+    // return, and flag for @DEFAULT node
+    return 0;
+  }
+  else if (strcmp(command,"EXIT") == 0 ) {
+    // return, and flag for global exit
+    return 1;
+  }
+
+  // do MD iterations until told to exit
+  while ( true ) {
+
+    // run an MD timestep
+    update->whichflag = 1; // 1 for dynamics
+    timer->init_timeout();
+    update->nsteps += 1;
+    update->laststep += 1;
+    update->endstep = update->laststep;
+    output->next = update->ntimestep + 1;
+    update->integrate->run(1);
+
+    // get the most recent command the MDI engine received
+    command = mdi_fix->command;
+
+    if (strcmp(command,"@DEFAULT") == 0 ) {
+      // return, and flag for @DEFAULT node
+      return 0;
+    }
+    else if (strcmp(command,"EXIT") == 0 ) {
+      // return, and flag for global exit
+      return 1;
+    }
+
+  }
+
+}
+
+
+
+int CommandMDIEngine::mdi_optg()
+{
+  char *command = NULL;
+
+  // create instance of the Minimizer class
+  Minimize *minimizer = new Minimize(lmp);
+
+  // initialize the minimizer in a way that ensures optimization will continue until MDI exits
+  update->etol = std::numeric_limits<double>::min();
+  update->ftol = std::numeric_limits<double>::min();
+  update->nsteps = std::numeric_limits<int>::max();
+  update->max_eval = std::numeric_limits<int>::max();
+
+  update->whichflag = 2; // 2 for minimization
+  update->beginstep = update->firststep = update->ntimestep;
+  update->endstep = update->laststep = update->firststep + update->nsteps;
+  lmp->init();
+
+  command = mdi_fix->engine_mode("@INIT_OPTG");
+
+  if (strcmp(command,"@DEFAULT") == 0 ) {
+    // return, and flag for @DEFAULT node
+    return 0;
+  }
+  else if (strcmp(command,"EXIT") == 0 ) {
+    // return, and flag for global exit
+    return 1;
+  }
+
+  update->minimize->setup();
+  command = mdi_fix->command;
+
+  if (strcmp(command,"@DEFAULT") == 0 ) {
+    // return, and flag for @DEFAULT node
+    return 0;
+  }
+  else if (strcmp(command,"EXIT") == 0 ) {
+    // return, and flag for global exit
+    return 1;
+  }
+
+  update->minimize->iterate(update->nsteps);
+  command = mdi_fix->command;
+
+  if (strcmp(command,"@DEFAULT") == 0 ) {
+    // return, and flag for @DEFAULT node
+    return 0;
+  }
+  else if (strcmp(command,"EXIT") == 0 ) {
+    // return, and flag for global exit
+    return 1;
+  }
+
+  error->all(FLERR,strcat("MDI reached end of OPTG simulation with invalid command: ",command));
```

Comment:
this is *really* bad and unsafe programming. You cannot just append a string to a constant string. This will likely cause segmentation faults. There is a much better solution for this using `fmt::format()`

Reply:
Thanks for catching this.  a1f957b switches to `fmt::format()`.